src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonDerivedType.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    /// <summary>
+    /// Represents a supported derived type defined in the metadata of a polymorphic type.
+    /// </summary>
+    public readonly struct JsonDerivedType
+    {
+        /// <summary>
+        /// Specifies a supported derived type without a type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        public JsonDerivedType(Type derivedType)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminator = null;
+        }
+
+        /// <summary>
+        /// Specifies a supported derived type with an integer type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        /// <param name="typeDiscriminator">The type discriminator to be associated with the derived type.</param>
+        public JsonDerivedType(Type derivedType, int typeDiscriminator)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminator = typeDiscriminator;
+        }
+
+        /// <summary>
+        /// Specifies a supported derived type with a string type discriminator.
+        /// </summary>
+        /// <param name="derivedType">The derived type to be supported by the polymorphic type metadata.</param>
+        /// <param name="typeDiscriminator">The type discriminator to be associated with the derived type.</param>
+        public JsonDerivedType(Type derivedType, string typeDiscriminator)
```

Comment:
should `typeDiscriminator` be nullable? If user writes `null` it won't make difference because we will pick `object?` overload but it will make a difference if string comes from variable like i.e. some external file

Reply:
type discriminators identifiers can't be null. `null` is special _internally_ in that it represents not using any type discriminators, but this shouldn't leak to public APIs.